xilem_web/src/event_handler.rs

```diff
@@ -0,0 +1,167 @@
+pub struct DeferEventHandler<State, Action, FOut, F, FF, CF> {
+    #[allow(clippy::complexity)]
+    phantom: PhantomData<fn() -> (State, Action, FOut, F)>,
+    future_fn: FF,
+    callback_fn: CF,
+}
+
+impl<State, Action, Event, FOut, F, FF, CF> EventHandler<Event, State, Action, ViewCtx>
+    for DeferEventHandler<State, Action, FOut, F, FF, CF>
+where
+    State: 'static,
+    Action: 'static,
+    Event: 'static,
+    FOut: Message,
+    F: Future<Output = FOut> + 'static,
+    FF: Fn(&mut State, Event) -> F + 'static,
+    CF: Fn(&mut State, FOut) -> Action + 'static,
+{
+    type State = Rc<MessageThunk>;
+
+    fn build(&self, ctx: &mut ViewCtx) -> Self::State {
+        Rc::new(ctx.message_thunk())
+    }
+
+    fn rebuild(&self, _prev: &Self, _event_handler_state: &mut Self::State, _ctx: &mut ViewCtx) {}
+
+    fn teardown(&self, _event_handler_state: &mut Self::State, _ctx: &mut ViewCtx) {}
+
+    fn message(
+        &self,
+        message_thunk: &mut Self::State,
+        id_path: &[xilem_core::ViewId],
+        message: EventHandlerMessage<Event>,
+        app_state: &mut State,
+    ) -> MessageResult<Action, EventHandlerMessage<Event>> {
+        debug_assert!(id_path.is_empty());
+        match message {
+            EventHandlerMessage::Event(event) => {
+                let future = (self.future_fn)(app_state, event);
+                let thunk = Rc::clone(message_thunk);
+                // TODO currently, multiple events could trigger this, while the (old) future is still not resolved
+                // This may be intended, but can also lead to surprising behavior.
+                // We could add an atomic boolean, to avoid this, i.e. either block a new future,
+                // or even queue it after the first future being resolved, there may also be other possible desired behaviors
+                // This could also be made configurable, e.g. via the builder pattern, like this:
+                // ```
+                // defer(...)
+                //     .block() // block new events triggering that future
+                //     .once() // allow this event to trigger the future only once.
+                //     .queue() // queue additional triggered futures
+                // ```
+                spawn_local(async move { thunk.push_message(future.await) });
+                MessageResult::RequestRebuild
+            }
+            EventHandlerMessage::Message(output) => MessageResult::Action((self.callback_fn)(
+                app_state,
+                *output.downcast::<FOut>().unwrap_throw(),
+            )),
+        }
+    }
+}
+
+pub fn defer<State, Action, Event, FOut, F, FF, CF>(
+    future_fn: FF,
+    callback_fn: CF,
+) -> DeferEventHandler<State, Action, FOut, F, FF, CF>
+where
+    State: 'static,
+    Action: 'static,
+    Event: 'static,
+    FOut: Message,
+    F: Future<Output = FOut> + 'static,
+    FF: Fn(&mut State, Event) -> F + 'static,
```

Comment:
I think this could also be 
```suggestion
    FF: Fn(&mut State, Event) -> Option<F> + 'static,
```

which would give the user more control, if they want to handle the event or not, thus maybe make the blocking/once stuff above a non issue, as the user could handle this with more control.

I think the parallel/serial(queue) thing explained in the comment above may still be useful though for configuration.

Reply:
But why would you call `defer` if you do not want to return a `Future`?

Reply:
... it would be like this, right?
```rust
|state, ev|{
  if state.value == some_value {
    None
  } else {
    Some(async { /* */ })
  }
}
```

Reply:
Yes